src/main/perl/Download/Retrieve.pm

```diff
@@ -0,0 +1,153 @@
+# ${license-info}
+# ${developer-info}
+# ${author-info}
+# ${build-info}
+
+package CAF::Download::Retrieve;
+
+use strict;
+use warnings;
+
+use Readonly;
+use LC::Exception qw (SUCCESS);
+
+Readonly my $MAX_RETRIES => 1000;
+
+=head1 NAME
+
+CAF::Download::Retrieve - Class for retrieval for L<CAF::Download>.
+
+=head1 DESCRIPTION
+
+This class handles the downloading of the URLs to use within L<CAF::Download>.
+
+=cut
+
+=head2 Functions
+
+=over
+
+=item C<prepare_destination>
+
+C<prepare_destination> prepares the destination.
+
+Returns the prepared destination on success, undef in case of failure (and sets the C<fail> attribute).
+No errors are logged.
+
+=cut
+
+# TODO: e.g. mkdir, check write rights, tempdir, in case there's no intermediate file...
+
+sub prepare_destination
+{
+    my ($self, $destination) = @_;
+
+    return $destination;
+}
+
+=pod
+
+=item download
+
+Download the data from the url(s) to the destination.
+In case a retrieval fails, the following url is tried. If there are no more urls to try,
+it will reiterate over the original list of urls, and this maximum C<retries> time per url,
+with a C<retry_wait> wait interval before each retry.
+
+Returns SUCCESS on succes, undef in case of failure (and sets the C<fail> attribute).
+No errors are logged.
+
+=cut
+
+sub download
+{
+    my ($self) = @_;
+
+    # in case the parse_urls failed. fail attribute is set
+    return if (!defined($self->{urls}));
+
+    my %tried; # per-url retry counter
+    my $tries = 0; # total tries
+
+    # a weak copy, so we can push/shift without changing original list of urls
+    my @urls = @{$self->{urls}};
+
+    while (@urls) {
+        my $url = shift @urls;
+
+        my $txt = $url->{_string};
+        my $id = $url->{_id};
+
+        $self->verbose("download url $txt (id $id) attempt ",
+                       ($tried{$id} || 0) + 1,
+                       " total attempts $tries.");
+
+        # TODO: should we really wait if we try another url?
```

Comment:
For a download that has N urls, they should be tried in sequence without delay, and only after all N have failed should there be any back off & wait. [IMHO]